Review

rshafikov opened the pull request:
final version before review

rshafikov committed:
fix: remove all redundant files

rshafikov committed:
feat: add config and base flags

(cherry picked from commit be5bcd6b128ac68eb7969437be36961cf8350635)

rshafikov committed:
fix: bump pgx version to fit GitHub Actions requirements

(cherry picked from commit 24de856b80217c0d1a8734c3511696dde408843a)

rshafikov committed:
fix: change an Env pointer and add DB URI check

(cherry picked from commit ff4ef2bf19b46b5a91940b06dc23d5e9b42cfd86)

rshafikov committed:
feat: add register and login endpoints to API

(cherry picked from commit c1abdc7ace72ee3a067db87ce4bb8569e5c10246)

rshafikov committed:
refactor: add graceful shutdown for server and small log fixes

(cherry picked from commit 27be1f5dcf06416d47098c5741856ecf6d2ab199)

rshafikov committed:
feat: add authentication middleware and secure order creation

(cherry picked from commit 7b3d0fdf75074dee895d668435868744bf506395)

rshafikov committed:
refactor: add tests for auth middleware, delete redundant packages

(cherry picked from commit 054925ea6999123d80b0af8d776916f70e403d13)

rshafikov committed:
refactor: change an authentication format, small fixes.

(cherry picked from commit 9c3513d5a353e4c05d07b0ef8215b1da6ab53002)

rshafikov committed:
fix: bump dependencies to fit static check

(cherry picked from commit dfd68754d88bdfad950c37a554d2aa260338077d)

rshafikov committed:
feat: add order management functionality

rshafikov committed:
feat: add database migrations and compress middleware

Implemented database migration functionality using `golang-migrate` to streamline schema updates. Added compression middleware to optimize server responses for JSON and plain text formats.

rshafikov committed:
fix: DATABASE_URI variable has default port 5432

rshafikov committed:
feat: add balance and transaction management

Added database schema for `balances` and `transactions` tables with relevant indexes. Implemented services, handlers, and repositories for user balance management. Integrated balance details and transaction history endpoints into the API. Updated order processing to handle balance adjustments.

rshafikov committed:
feat: implement withdrawal handling and update balance logic

Added withdrawal schema and integrated it into the API for handling user withdrawals. Adjusted balance and transaction services to support withdrawal operations. Enhanced error handling for invalid input, insufficient funds, and updated balance repository with transaction creation.

rshafikov committed:
refactor: improve response handling and update task configuration

Enhanced response handling across handlers by removing redundant `WriteHeader` calls and properly setting headers. Updated `Taskfile.yml` to include `DATABASE_URI` for testing and adjusted default ports in CI/CD tasks. Added middleware to enforce consistent `Content-Type` headers.

rshafikov committed:
refactor: replace transactions with withdrawals, adjust logic and handlers

Replaced the `transactions` structure and implementation with `withdrawals` across the codebase. Updated database schema, repositories, services, and handlers to reflect this change. Enhanced withdrawal validation, response handling, and balance management logic to improve reliability.

rshafikov committed:
refactor: replace `Withdraw` schema with comprehensive withdrawal handling tests

Removed the unused `Withdraw` schema. Added extensive tests for balance-related operations, including fetching user balance, withdrawals, and executing withdrawal requests. Enhanced mocks and handlers to support detailed testing of withdrawal logic.

rshafikov committed:
feat: introduce worker pool for order processing and balance updates

Added a worker pool implementation for handling order processing tasks in parallel. Integrated it into the router and enhanced order handling to assign tasks to the pool. Updated balance service to support incremental balance updates.

rshafikov committed:
Merge branch 'master' into review

ekubyshin commented on internal/app/app.go:
```diff
@@ -29,29 +31,47 @@ func NewApplication(cfg defaultConfig) *Application {
 }
 
-func (app *Application) ConnectToDatabase(ctx context.Context) error {
+func (app *Application) ConnectToDatabase(ctx context.Context) {
```
подобные вещи обычно выносят в отдельный пакет db.go, а потом вызывают просто конструктор, который возвращает готовый коннект, чтобы сделать декомпозицию.

ekubyshin commented on internal/app/app.go:
```diff
@@ -29,29 +31,47 @@ func NewApplication(cfg defaultConfig) *Application {
+
+}
+
+func (app *Application) MigrateDatabase(ctx context.Context) {
```
при желании можно вынести в отельный файл migrate.go и запускать из github action, для этого можно отредактировать yml файл гитхаба. Тк в проектах обычно принято разделять миграцию от запуска, чтобы была возможность откатить ее или не раскатывать приложение, если миграция не проходит.

ekubyshin commented on internal/client/client.go:
отлично что вынес отдельного клиента в пакет и определил ошибки!

ekubyshin commented on internal/database/queries/wd_query.go:
хороший подход к организации работы с sql запросами

ekubyshin commented on internal/handlers/balance_handler.go:
```diff
@@ -0,0 +1,123 @@
+	BalanceService models.BalanceService
+}
+
+func NewBalanceHandler(service models.BalanceService) *BalanceHandler {
```
круто, что выделил слой бизнес логики. Интерфейс можно объявить прям тут перед конструктором, что будет соответствовать идеоматическому подходу в го

ekubyshin commented on internal/handlers/order_handler.go:
```diff
@@ -0,0 +1,109 @@
+	WP      *workerpool.WorkerPool
+}
+
+func NewOrderHandler(orderService models.OrderService, pool *workerpool.WorkerPool) *OrderHandler {
```
отличная декомпозиция хэндлеров

ekubyshin commented on internal/handlers/user_handler.go:
```diff
@@ -38,7 +36,7 @@ func (h *UserHandler) Register(w http.ResponseWriter, r *http.Request) {
 	}
 
-	if err := h.UserService.Register(ctx, reqUser.Login, reqUser.Password); err != nil {
+	if err := h.UserService.Register(r.Context(), reqUser.Login, reqUser.Password); err != nil {
```
контекст передан верно

rshafikov committed:
refactor: replace model service interfaces with lightweight contract interfaces

Simplified service interface definitions by introducing lightweight contract interfaces localized to their use cases. Updated constructors, handlers, and worker pool to use new interfaces, reducing coupling and improving testability.

rshafikov committed:
refactor: simplify database setup and remove redundant migration logic

Streamlined the database setup process by combining connection and migration logic into `SetupDB`. Removed unused `ConnectToDatabase` and `MigrateDatabase` methods. Updated the application to use the new workflow and removed unnecessary SQL migration files.

rshafikov committed:
refactor: simplify database setup and remove redundant migration logic

Streamlined the database setup process by combining connection and migration logic into `SetupDB`. Removed unused `ConnectToDatabase` and `MigrateDatabase` methods. Updated the application to use the new workflow and removed unnecessary SQL migration files.

rshafikov committed:
refactor: update mocks and tests for service interface changes

Regenerated GoMock files to align with updated service interfaces. Updated affected tests and handlers to use the new mock implementation and naming conventions.
